src/main/java/io/odpf/firehose/sink/clickhouse/QueryTemplate.java

```diff
@@ -0,0 +1,118 @@
+        template = Mustache.compiler().withEscaper(Escapers.simple()).compile(query);
+    }
+
+    private void initialize(ClickhouseSinkConfig clickhouseSinkConfig) {
+        scopes.put("table", clickhouseSinkConfig.getClickhouseTableName());
+
+        Properties messageProtoToDBColumnsMapping = clickhouseSinkConfig.getInputSchemaProtoToColumnMapping();
+        insertColumns = getInsertColumns(messageProtoToDBColumnsMapping);
+        scopes.put("insertColumns", String.join(",", insertColumns));
+    }
+
+    private List<String> getInsertColumns(Properties messageProtoToDBColumnsMapping) {
+        List<String> columns = new ArrayList<>();
+        Enumeration<?> propertyNames = messageProtoToDBColumnsMapping.propertyNames();
+        while (propertyNames.hasMoreElements()) {
+            Object tableColumn = messageProtoToDBColumnsMapping.get(propertyNames.nextElement());
+            if (tableColumn instanceof String) {
+                columns.add((String) tableColumn);
+            } else if (tableColumn instanceof Properties) {
+                columns.addAll(getInsertColumns((Properties) tableColumn));
+            }
+        }
+        return columns;
+    }
+
+    public String toQueryStringForSingleMessage(Message message) {
+
+        byte[] value;
+
+        if ("message".equals(kafkaRecordParserMode)) {
+            value = message.getLogMessage();
+        } else {
+            value = message.getLogKey();
+        }
+
+        Map<String, Object> columnToValue = protoToFieldMapper.getFields(value);
+
+        String insertValues = stringifyColumnValues(columnToValue, insertColumns);
+
+        scopes.put("insertValues", insertValues);
+
+        return template.execute(scopes);
+    }
+
+    private String stringifyColumnValues(Map<String, Object> columnToValue, List<String> columns) {
+        List<String> columnValues = columns.stream()
+                .map(c -> columnToValue.get(c).toString().replace("'", "''"))
+                .map(c -> "\'" + c + "\'")
+                .collect(Collectors.toList());
+        return String.join(", ", columnValues);
+    }
+
+    public String toQueryStringForMultipleMessages(List<Message> messages) {
+        byte[] value;
+        String insertValues = "";
+
+        long count = 0; //counter to maintain number of rows in the query.
+        for (Message message : messages) {
+            if ("message".equals(kafkaRecordParserMode)) {
+                value = message.getLogMessage();
+            } else {
+                value = message.getLogKey();
+            }
+            count++;
+            Map<String, Object> columnToValue = protoToFieldMapper.getFields(value);
+            insertValues = insertValues + stringifyColumnValues(columnToValue, insertColumns);
+
+            /*
+            Value seperator is not needed if it's the last row to be added.
+             */
+            if (count != messages.size()) {
+                insertValues = insertValues + VALUES_SEPERATOR;
```

Comment:
instead of this use can use String.join(VALUES_SEPERATOR,list)  after the loop.
https://docs.oracle.com/javase/8/docs/api/java/lang/String.html#join-java.lang.CharSequence-java.lang.CharSequence...-
or better you can use 

```
insertValues = messages.map(m -> {// do the transform logic}).collectAsList());
scopes.put("insertvalues", String.join(VALUES_SEPERATOR, insertValues);

```